// mxw01-thermal-printer/dist/index.js
const I = [
  0,
  7,
  14,
  9,
  28,
  27,
  18,
  21,
  56,
  63,
  54,
  49,
  36,
  35,
  42,
  45,
  112,
  119,
  126,
  121,
  108,
  107,
  98,
  101,
  72,
  79,
  70,
  65,
  84,
  83,
  90,
  93,
  224,
  231,
  238,
  233,
  252,
  251,
  242,
  245,
  216,
  223,
  214,
  209,
  196,
  195,
  202,
  205,
  144,
  151,
  158,
  153,
  140,
  139,
  130,
  133,
  168,
  175,
  166,
  161,
  180,
  179,
  186,
  189,
  199,
  192,
  201,
  206,
  219,
  220,
  213,
  210,
  255,
  248,
  241,
  246,
  227,
  228,
  237,
  234,
  183,
  176,
  185,
  190,
  171,
  172,
  165,
  162,
  143,
  136,
  129,
  134,
  147,
  148,
  157,
  154,
  39,
  32,
  41,
  46,
  59,
  60,
  53,
  50,
  31,
  24,
  17,
  22,
  3,
  4,
  13,
  10,
  87,
  80,
  89,
  94,
  75,
  76,
  69,
  66,
  111,
  104,
  97,
  102,
  115,
  116,
  125,
  122,
  137,
  142,
  135,
  128,
  149,
  146,
  155,
  156,
  177,
  182,
  191,
  184,
  173,
  170,
  163,
  164,
  249,
  254,
  247,
  240,
  229,
  226,
  235,
  236,
  193,
  198,
  207,
  200,
  221,
  218,
  211,
  212,
  105,
  110,
  103,
  96,
  117,
  114,
  123,
  124,
  81,
  86,
  95,
  88,
  77,
  74,
  67,
  68,
  25,
  30,
  23,
  16,
  5,
  2,
  11,
  12,
  33,
  38,
  47,
  40,
  61,
  58,
  51,
  52,
  78,
  73,
  64,
  71,
  82,
  85,
  92,
  91,
  118,
  113,
  120,
  127,
  106,
  109,
  100,
  99,
  62,
  57,
  48,
  55,
  34,
  37,
  44,
  43,
  6,
  1,
  8,
  15,
  26,
  29,
  20,
  19,
  174,
  169,
  160,
  167,
  178,
  181,
  188,
  187,
  150,
  145,
  152,
  159,
  138,
  141,
  132,
  131,
  222,
  217,
  208,
  215,
  194,
  197,
  204,
  203,
  230,
  225,
  232,
  239,
  250,
  253,
  244,
  243
];
function D(a) {
  let t = 0;
  for (const e of a)
    t = I[(t ^ e) & 255];
  return t & 255;
}
function v(a) {
  return new Promise((t) => setTimeout(() => t(), a));
}
const d = {
  GetStatus: 161,
  SetIntensity: 162,
  PrintRequest: 169,
  FlushData: 173,
  PrintComplete: 170
}, g = {
  HEADER_BYTE_1: 34,
  HEADER_BYTE_2: 33,
  TERMINATOR: 255
};
function b(a, t) {
  const e = t.length, r = new Uint8Array([
    g.HEADER_BYTE_1,
    g.HEADER_BYTE_2,
    a,
    0,
    e & 255,
    e >> 8 & 255
  ]), i = new Uint8Array(r.length + t.length);
  i.set(r), i.set(t, r.length);
  const s = D(t), n = new Uint8Array(i.length + 2);
  return n.set(i), n[n.length - 2] = s, n[n.length - 1] = g.TERMINATOR, n;
}
function N(a) {
  if (a[0] !== g.HEADER_BYTE_1 || a[1] !== g.HEADER_BYTE_2)
    return null;
  const t = a[2], e = a[4] | a[5] << 8, r = a.slice(6, 6 + e);
  return { cmdId: t, payload: r };
}
function M(a) {
  if (a.length < 7)
    return null;
  const t = a[6];
  return {
    printing: (t & 1) !== 0,
    paper_jam: (t & 2) !== 0,
    out_of_paper: (t & 4) !== 0,
    cover_open: (t & 8) !== 0,
    battery_low: (t & 16) !== 0,
    overheat: (t & 32) !== 0
  };
}
function L() {
  return {
    printing: !1,
    paper_jam: !1,
    out_of_paper: !1,
    cover_open: !1,
    battery_low: !1,
    overheat: !1
  };
}
class O {
  state;
  printComplete = !1;
  pendingResolvers = /* @__PURE__ */ new Map();
  constructor() {
    this.state = L();
  }
  /**
   * Get current printer state
   */
  getState() {
    return { ...this.state };
  }
  /**
   * Check if print is complete
   */
  isPrintComplete() {
    return this.printComplete;
  }
  /**
   * Reset print complete flag
   */
  resetPrintComplete() {
    this.printComplete = !1;
  }
  /**
   * Process notification and update state
   * @param cmdId Command ID from notification
   * @param payload Notification payload
   */
  processNotification(t, e) {
    if (t === d.PrintComplete && (this.printComplete = !0), t === d.GetStatus) {
      const i = M(e);
      i && (this.state = i);
    }
    const r = this.pendingResolvers.get(t);
    r && (r(e), this.pendingResolvers.delete(t));
  }
  /**
   * Wait for notification response
   * @param cmdId Command ID to wait for
   * @param timeoutMs Timeout in milliseconds
   * @returns Promise that resolves with the payload
   */
  waitForNotification(t, e = 1e4) {
    return new Promise((r, i) => {
      const s = setTimeout(() => {
        this.pendingResolvers.delete(t), i(
          new Error(`Timeout waiting for notification 0x${t.toString(16)}`)
        );
      }, e);
      this.pendingResolvers.set(t, (n) => {
        clearTimeout(s), r(n);
      });
    });
  }
}
const y = 384, m = y / 8, S = 90 * m;
class W {
  controlWrite;
  dataWrite;
  stateManager;
  constructor(t, e) {
    this.controlWrite = t, this.dataWrite = e, this.stateManager = new O();
  }
  /**
   * Get current printer state
   */
  get state() {
    return this.stateManager.getState();
  }
  /**
   * Process incoming notification from printer
   */
  notify(t) {
    const e = N(t);
    if (!e) {
      console.warn("Ignoring unexpected notification format");
      return;
    }
    this.stateManager.processNotification(e.cmdId, e.payload);
  }
  /**
   * Set print intensity (darkness)
   */
  async setIntensity(t = 93) {
    const e = b(d.SetIntensity, Uint8Array.of(t));
    await this.controlWrite(e), await v(50);
  }
  /**
   * Request current printer status
   */
  async requestStatus() {
    const t = b(d.GetStatus, Uint8Array.of(0));
    return await this.controlWrite(t), this.stateManager.waitForNotification(d.GetStatus, 5e3);
  }
  /**
   * Send print request with number of lines
   */
  async printRequest(t, e = 0) {
    const r = new Uint8Array(4);
    r[0] = t & 255, r[1] = t >> 8 & 255, r[2] = 48, r[3] = e;
    const i = b(d.PrintRequest, r);
    return await this.controlWrite(i), this.stateManager.waitForNotification(d.PrintRequest, 5e3);
  }
  /**
   * Flush data to printer
   */
  async flushData() {
    const t = b(d.FlushData, Uint8Array.of(0));
    await this.controlWrite(t), await v(50);
  }
  /**
   * Send data chunks to printer
   */
  async sendDataChunks(t, e = m) {
    let r = 0;
    for (; r < t.length; ) {
      const i = t.slice(r, Math.min(r + e, t.length));
      await this.dataWrite(i), r += i.length, await v(15);
    }
  }
  /**
   * Wait for print completion
   */
  async waitForPrintComplete(t = 2e4) {
    this.stateManager.resetPrintComplete();
    const e = Date.now();
    for (; !this.stateManager.isPrintComplete() && Date.now() - e < t; )
      await v(100);
    if (!this.stateManager.isPrintComplete())
      throw new Error("Print timeout: Did not receive completion notification");
  }
}
function x(a) {
  if (a.length !== y)
    throw new Error(
      `Row length must be ${y}, got ${a.length}`
    );
  const t = new Uint8Array(m);
  for (let e = 0; e < m; e++) {
    let r = 0;
    for (let i = 0; i < 8; i++)
      a[e * 8 + i] && (r |= 1 << i);
    t[e] = r;
  }
  return t;
}
function B(a) {
  const t = a.length;
  let e = new Uint8Array(0);
  for (let r = 0; r < t; r++) {
    const i = x(a[r]), s = new Uint8Array(e.length + i.length);
    s.set(e), s.set(i, e.length), e = s;
  }
  if (e.length < S) {
    const r = new Uint8Array(S - e.length), i = new Uint8Array(e.length + r.length);
    i.set(e), i.set(r, e.length), e = i;
  }
  return e;
}
class U {
  listeners = /* @__PURE__ */ new Map();
  /**
   * Subscribe to an event
   * @param eventType Event type to listen for
   * @param listener Callback function
   * @returns Unsubscribe function
   */
  on(t, e) {
    return this.listeners.has(t) || this.listeners.set(t, /* @__PURE__ */ new Set()), this.listeners.get(t).add(e), () => {
      const r = this.listeners.get(t);
      r && r.delete(e);
    };
  }
  /**
   * Emit an event to all listeners
   * @param event Event to emit
   */
  emit(t) {
    const e = this.listeners.get(t.type);
    e && e.forEach((r) => {
      try {
        r(t);
      } catch (i) {
        console.error("Error in event listener:", i);
      }
    });
  }
  /**
   * Clear all event listeners
   */
  clear() {
    this.listeners.clear();
  }
  /**
   * Remove all listeners for a specific event type
   * @param eventType Event type to clear
   */
  clearEventType(t) {
    this.listeners.delete(t);
  }
}
class F {
  _isConnected = !1;
  _isPrinting = !1;
  _printerState = null;
  _statusMessage = "Ready to connect printer";
  _ditherMethod = "steinberg";
  _printIntensity = 93;
  // Getters
  get isConnected() {
    return this._isConnected;
  }
  get isPrinting() {
    return this._isPrinting;
  }
  get printerState() {
    return this._printerState;
  }
  get statusMessage() {
    return this._statusMessage;
  }
  get ditherMethod() {
    return this._ditherMethod;
  }
  get printIntensity() {
    return this._printIntensity;
  }
  // Setters with validation
  setConnected(t) {
    this._isConnected = t;
  }
  setPrinting(t) {
    this._isPrinting = t;
  }
  setPrinterState(t) {
    this._printerState = t;
  }
  setStatusMessage(t) {
    this._statusMessage = t;
  }
  setDitherMethod(t) {
    this._ditherMethod = t;
  }
  setPrintIntensity(t) {
    if (t < 0 || t > 255)
      throw new Error("Print intensity must be between 0 and 255");
    this._printIntensity = t;
  }
  /**
   * Reset to initial state
   */
  reset() {
    this._isConnected = !1, this._isPrinting = !1, this._printerState = null, this._statusMessage = "Ready to connect printer";
  }
}
class w {
}
class H extends w {
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  apply(t, e, r) {
    for (let i = 0; i < t.length; ++i)
      t[i] = t[i] > 128 ? 255 : 0;
    return t;
  }
  getName() {
    return "threshold";
  }
}
class V extends w {
  apply(t, e, r) {
    let i = 0;
    for (let s = 0; s < r; ++s)
      for (let n = 0; n < e; ++n) {
        const c = t[i], o = c > 128 ? 255 : 0, h = c - o;
        t[i] = o, n < e - 1 && (t[i + 1] += h * 7 / 16), s < r - 1 && (n > 0 && (t[i + e - 1] += h * 3 / 16), t[i + e] += h * 5 / 16, n < e - 1 && (t[i + e + 1] += h / 16)), ++i;
      }
    return t;
  }
  getName() {
    return "steinberg";
  }
}
class q extends w {
  bayer8 = [
    0,
    48,
    12,
    60,
    3,
    51,
    15,
    63,
    32,
    16,
    44,
    28,
    35,
    19,
    47,
    31,
    8,
    56,
    4,
    52,
    11,
    59,
    7,
    55,
    40,
    24,
    36,
    20,
    43,
    27,
    39,
    23,
    2,
    50,
    14,
    62,
    1,
    49,
    13,
    61,
    34,
    18,
    46,
    30,
    33,
    17,
    45,
    29,
    10,
    58,
    6,
    54,
    9,
    57,
    5,
    53,
    42,
    26,
    38,
    22,
    41,
    25,
    37,
    21
  ];
  ditherFactor = 0.6;
  apply(t, e, r) {
    let i = 0;
    for (let s = 0; s < r; ++s)
      for (let n = 0; n < e; ++n) {
        const c = this.bayer8[s % 8 * 8 + n % 8];
        let o = t[i];
        o = o + (c - 32) * this.ditherFactor, o = Math.max(0, Math.min(255, o)), t[i] = o > 128 ? 255 : 0, ++i;
      }
    return t;
  }
  getName() {
    return "bayer";
  }
}
class Y extends w {
  apply(t, e, r) {
    let i = 0;
    for (let s = 0; s < r; ++s)
      for (let n = 0; n < e; ++n) {
        const c = t[i], o = c > 128 ? 255 : 0, h = c - o >> 3;
        t[i] = o, n < e - 1 && (t[i + 1] += h), n < e - 2 && (t[i + 2] += h), s < r - 1 && (n > 0 && (t[i + e - 1] += h), t[i + e] += h, n < e - 1 && (t[i + e + 1] += h)), s < r - 2 && (t[i + 2 * e] += h), ++i;
      }
    return t;
  }
  getName() {
    return "atkinson";
  }
}
class j extends w {
  apply(t, e, r) {
    for (let c = 0; c < r - 4; c += 4) {
      for (let o = 0; o < e - 4; o += 4) {
        let h = 0;
        for (let l = 0; l < 4; ++l)
          for (let p = 0; p < 4; ++p)
            h += t[(c + p) * e + o + l];
        const f = (1 - h / 16 / 255) * 4;
        for (let l = 0; l < 4; ++l)
          for (let p = 0; p < 4; ++p)
            t[(c + p) * e + o + l] = Math.abs(l - 3) >= f || Math.abs(p - 3) >= f ? 255 : 0;
      }
      for (let o = e - e % 4; o < e; ++o)
        t[c * e + o] = 255;
    }
    for (let c = r - r % 4; c < r; ++c)
      for (let o = 0; o < e; ++o)
        t[c * e + o] = 255;
    return t;
  }
  getName() {
    return "pattern";
  }
}
function k(a) {
  switch (a) {
    case "steinberg":
      return new V();
    case "bayer":
      return new q();
    case "atkinson":
      return new Y();
    case "pattern":
      return new j();
    case "threshold":
    default:
      return new H();
  }
}
function $(a, t, e, r) {
  if (r === 0)
    return a;
  const i = new Uint8ClampedArray(a.length);
  switch (r) {
    case 90:
      for (let s = 0; s < e; s++)
        for (let n = 0; n < t; n++) {
          const c = e - 1 - s, o = n;
          i[o * e + c] = a[s * t + n];
        }
      break;
    case 180:
      for (let s = 0; s < e; s++)
        for (let n = 0; n < t; n++) {
          const c = t - 1 - n, o = e - 1 - s;
          i[o * t + c] = a[s * t + n];
        }
      break;
    case 270:
      for (let s = 0; s < e; s++)
        for (let n = 0; n < t; n++) {
          const c = s, o = t - 1 - n;
          i[o * e + c] = a[s * t + n];
        }
      break;
  }
  return i;
}
function G(a, t, e, r) {
  if (r === "none")
    return a;
  const i = new Uint8ClampedArray(a.length);
  switch (r) {
    case "h":
      for (let s = 0; s < e; s++)
        for (let n = 0; n < t; n++)
          i[s * t + n] = a[s * t + (t - n - 1)];
      break;
    case "v":
      for (let s = 0; s < e; s++)
        for (let n = 0; n < t; n++)
          i[s * t + n] = a[(e - s - 1) * t + n];
      break;
    case "both":
      for (let s = 0; s < e; s++)
        for (let n = 0; n < t; n++)
          i[s * t + n] = a[(e - s - 1) * t + (t - n - 1)];
      break;
  }
  return i;
}
function X(a, t = 128, e = !0) {
  const r = new Uint8ClampedArray(a.length);
  for (let i = 0; i < r.length; ++i) {
    const s = a[i];
    let n = s & 255, c = s >> 8 & 255, o = s >> 16 & 255;
    const h = (s >> 24 & 255) / 255;
    if (h < 1 && e) {
      const l = 1 - h;
      n += (255 - n) * l, c += (255 - c) * l, o += (255 - o) * l;
    } else
      n *= h, c *= h, o *= h;
    let f = n * 0.2125 + c * 0.7154 + o * 0.0721;
    f += (t - 128) * (1 - f / 255) * (f / 255) * 2, r[i] = f;
  }
  return r;
}
function J(a, t = !1) {
  const e = new Uint32Array(a.length);
  for (let r = 0; r < a.length; ++r) {
    const i = a[r] === 255 && t ? 0 : 4278190080;
    e[r] = i | a[r] << 16 | a[r] << 8 | a[r];
  }
  return e;
}
function z(a, t, e) {
  const r = Math.min(t, a.width), i = Math.min(e, a.height), s = new Uint8ClampedArray(r * i * 4);
  for (let n = 0; n < i; n++)
    for (let c = 0; c < r; c++) {
      const o = (n * a.width + c) * 4, h = (n * r + c) * 4;
      s[h] = a.data[o], s[h + 1] = a.data[o + 1], s[h + 2] = a.data[o + 2], s[h + 3] = a.data[o + 3];
    }
  return {
    data: s,
    width: r,
    height: i
  };
}
function K(a, t) {
  const e = new Uint32Array(
    new Uint8ClampedArray(a.data).buffer
  ), r = a.width, i = a.height;
  let s = X(e, t.brightness, !0);
  s = k(t.dither).apply(s, r, i), s = G(s, r, i, t.flip);
  let c = r, o = i;
  s = $(s, r, i, t.rotate), (t.rotate === 90 || t.rotate === 270) && (c = i, o = r);
  const h = J(s, !0), f = 384, l = [];
  for (let p = 0; p < o; p++) {
    const E = [];
    for (let C = 0; C < c; C++) {
      const T = p * c + C, A = s[T];
      E.push(A < 128);
    }
    for (; E.length < f; )
      E.push(!1);
    l.push(E);
  }
  return {
    processedData: h,
    width: c,
    height: o,
    binaryRows: l
  };
}
class Q {
  imageData;
  options;
  constructor(t, e = {}) {
    this.imageData = t, this.options = e;
  }
  /**
   * Process and prepare image for printing
   * @param defaultDither Default dithering method
   * @returns Prepared image buffer and metadata
   */
  prepare(t) {
    const e = {
      dither: this.options.dither ?? t,
      brightness: this.options.brightness ?? 128,
      flip: this.options.flip ?? "none",
      rotate: this.options.rotate ?? 0
    };
    let r = this.imageData;
    this.imageData.width > y && (r = z(
      this.imageData,
      y,
      this.imageData.height
    ));
    const { binaryRows: i } = K(
      r,
      e
    );
    return {
      imageBuffer: B(i),
      numLines: i.length
    };
  }
  /**
   * Get print intensity from options or default
   * @param defaultIntensity Default intensity value
   * @returns Print intensity
   */
  getIntensity(t) {
    return this.options.intensity ?? t;
  }
}
class Z {
  adapter;
  printer = null;
  connection = null;
  device = null;
  eventEmitter;
  state;
  constructor(t) {
    if (!t.isAvailable())
      throw new Error("Bluetooth is not available in this environment");
    this.adapter = t, this.eventEmitter = new U(), this.state = new F();
  }
  // Public getters delegated to state
  get isConnected() {
    return this.state.isConnected;
  }
  get isPrinting() {
    return this.state.isPrinting;
  }
  get printerState() {
    return this.state.printerState;
  }
  get statusMessage() {
    return this.state.statusMessage;
  }
  get ditherMethod() {
    return this.state.ditherMethod;
  }
  get printIntensity() {
    return this.state.printIntensity;
  }
  // Public setters delegated to state
  setDitherMethod(t) {
    this.state.setDitherMethod(t);
  }
  setPrintIntensity(t) {
    this.state.setPrintIntensity(t);
  }
  /**
   * Subscribe to events
   */
  on(t, e) {
    return this.eventEmitter.on(t, e);
  }
  /**
   * Update status message and emit state change if needed
   */
  updateStatus(t, e) {
    this.state.setStatusMessage(t), e && (this.state.setPrinterState(e), this.eventEmitter.emit({ type: "stateChange", state: e }));
  }
  /**
   * Connect to printer via Bluetooth
   */
  async connect() {
    try {
      this.updateStatus("Connecting to printer..."), this.device = await this.adapter.requestDevice(), this.connection = await this.adapter.connect(this.device), this.printer = new W(
        this.connection.controlCharacteristic.writeValueWithoutResponse.bind(
          this.connection.controlCharacteristic
        ),
        this.connection.dataCharacteristic.writeValueWithoutResponse.bind(
          this.connection.dataCharacteristic
        )
      ), await this.setupNotifications(), this.state.setConnected(!0), this.updateStatus("Printer connected"), this.eventEmitter.emit({ type: "connected", device: this.device }), await this.getStatus();
    } catch (t) {
      const e = t;
      throw this.updateStatus(`Error: ${e.message}`), this.eventEmitter.emit({ type: "error", error: e }), t;
    }
  }
  /**
   * Setup notification listener
   */
  async setupNotifications() {
    if (!this.connection || !this.printer)
      throw new Error("No connection or printer available");
    const t = (e) => {
      const i = e.target.value;
      i && this.printer && (this.printer.notify(new Uint8Array(i.buffer)), this.updateStatus("Printer state updated", { ...this.printer.state }));
    };
    await this.connection.notifyCharacteristic.startNotifications(), this.connection.notifyCharacteristic.addEventListener(
      "characteristicvaluechanged",
      t
    );
  }
  /**
   * Get current printer status
   */
  async getStatus() {
    if (!this.printer || !this.state.isConnected)
      return this.updateStatus("Printer not connected"), null;
    try {
      await this.printer.requestStatus();
      const t = { ...this.printer.state };
      return this.updateStatus("Status updated", t), t;
    } catch (t) {
      const e = t;
      return this.updateStatus(`Error: ${e.message}`), this.eventEmitter.emit({ type: "error", error: e }), null;
    }
  }
  /**
   * Print from image data
   */
  async print(t, e = {}) {
    if (!this.printer || !this.state.isConnected)
      throw new Error("Printer not connected");
    try {
      this.state.setPrinting(!0), this.updateStatus("Preparing to print...");
      const r = new Q(t, e), { imageBuffer: i, numLines: s } = r.prepare(
        this.state.ditherMethod
      ), n = r.getIntensity(this.state.printIntensity);
      this.updateStatus("Configuring printer..."), await this.printer.setIntensity(n);
      const c = await this.printer.requestStatus();
      if (c.length >= 13 && c[12] !== 0)
        throw new Error(`Printer error: ${c[13]}`);
      this.updateStatus("Sending data...");
      const o = await this.printer.printRequest(s, 0);
      if (!o || o[0] !== 0)
        throw new Error("Print request rejected");
      await this.printer.sendDataChunks(i), await this.printer.flushData(), this.updateStatus("Printing..."), await this.printer.waitForPrintComplete(), this.updateStatus("Print completed"), await this.getStatus();
    } catch (r) {
      const i = r;
      throw this.updateStatus(`Error: ${i.message}`), this.eventEmitter.emit({ type: "error", error: i }), r;
    } finally {
      this.state.setPrinting(!1);
    }
  }
  /**
   * Disconnect from printer
   */
  async disconnect() {
    if (this.connection?.notifyCharacteristic)
      try {
        await this.connection.notifyCharacteristic.stopNotifications();
      } catch (t) {
        console.warn("Error stopping notifications:", t);
      }
    if (this.connection)
      try {
        await this.connection.disconnect();
      } catch (t) {
        console.warn("Error disconnecting:", t);
      }
    this.printer = null, this.connection = null, this.device = null, this.state.reset(), this.updateStatus("Printer disconnected"), this.eventEmitter.emit({ type: "disconnected" });
  }
  /**
   * Dispose of the client and clean up resources
   */
  dispose() {
    this.disconnect(), this.eventEmitter.clear();
  }
}
const u = {
  // Service UUIDs
  PRINTER_SERVICE: "0000ae30-0000-1000-8000-00805f9b34fb",
  PRINTER_SERVICE_ALT: "0000af30-0000-1000-8000-00805f9b34fb",
  // macOS alternate
  // Characteristic UUIDs
  CONTROL: "0000ae01-0000-1000-8000-00805f9b34fb",
  NOTIFY: "0000ae02-0000-1000-8000-00805f9b34fb",
  DATA: "0000ae03-0000-1000-8000-00805f9b34fb",
  // Short format for Noble (Node.js)
  CONTROL_SHORT: "ae01",
  NOTIFY_SHORT: "ae02",
  DATA_SHORT: "ae03"
};
class P {
  dataListeners = /* @__PURE__ */ new Map();
  /**
   * Clear all registered listeners
   */
  clearListeners() {
    this.dataListeners.clear();
  }
}
class _ extends P {
  constructor(t) {
    super(), this.characteristic = t;
  }
  async writeValueWithoutResponse(t) {
    await this.characteristic.writeValueWithoutResponse(t);
  }
  async startNotifications() {
    await this.characteristic.startNotifications();
  }
  async stopNotifications() {
    await this.characteristic.stopNotifications();
  }
  addEventListener(t, e) {
    this.characteristic.addEventListener(t, e);
  }
  removeEventListener(t, e) {
    this.characteristic.removeEventListener(t, e);
  }
}
class tt {
  device = null;
  server = null;
  /**
   * Check if Web Bluetooth is available
   */
  isAvailable() {
    return typeof navigator < "u" && typeof navigator.bluetooth < "u";
  }
  /**
   * Request a Bluetooth device with printer services
   */
  async requestDevice() {
    if (!this.isAvailable())
      throw new Error("Web Bluetooth API is not available in this browser");
    try {
      return this.device = await navigator.bluetooth.requestDevice({
        filters: [
          { services: [u.PRINTER_SERVICE] },
          { services: [u.PRINTER_SERVICE_ALT] }
        ],
        optionalServices: [
          u.PRINTER_SERVICE,
          u.PRINTER_SERVICE_ALT
        ]
      }), {
        id: this.device.id,
        name: this.device.name
      };
    } catch (t) {
      throw new Error(
        `Failed to request Bluetooth device: ${t.message}`
      );
    }
  }
  /**
   * Connect to a Bluetooth device and get service characteristics
   */
  async connect(t) {
    if (!this.device || this.device.id !== t.id)
      throw new Error("Device not found. Please request device first.");
    try {
      const e = this.device.gatt;
      if (!e)
        throw new Error("GATT not available on device");
      if (this.server = await e.connect(), !this.server)
        throw new Error("Failed to connect to GATT server");
      let r;
      try {
        r = await this.server.getPrimaryService(
          u.PRINTER_SERVICE
        );
      } catch {
        console.log("Trying alternate UUID for macOS compatibility..."), r = await this.server.getPrimaryService(
          u.PRINTER_SERVICE_ALT
        );
      }
      const [i, s, n] = await Promise.all([
        r.getCharacteristic(u.CONTROL),
        r.getCharacteristic(u.NOTIFY),
        r.getCharacteristic(u.DATA)
      ]);
      return {
        device: t,
        disconnect: async () => {
          this.server?.connected && this.server.disconnect(), this.device = null, this.server = null;
        },
        controlCharacteristic: new _(
          i
        ),
        dataCharacteristic: new _(n),
        notifyCharacteristic: new _(s)
      };
    } catch (e) {
      throw new Error(
        `Failed to connect to device: ${e.message}`
      );
    }
  }
}
class R extends P {
  characteristic;
  constructor(t) {
    super(), this.characteristic = t;
  }
  async writeValueWithoutResponse(t) {
    const e = Buffer.from(t);
    await this.characteristic.writeAsync(e, !0);
  }
  async startNotifications() {
    await this.characteristic.subscribeAsync();
  }
  async stopNotifications() {
    await this.characteristic.unsubscribeAsync();
  }
  addEventListener(t, e) {
    if (t === "characteristicvaluechanged") {
      const r = (i) => {
        e({
          target: {
            value: {
              buffer: i.buffer.slice(
                i.byteOffset,
                i.byteOffset + i.byteLength
              )
            }
          }
        });
      };
      this.dataListeners.set(e, r), this.characteristic.on("data", r);
    }
  }
  removeEventListener(t, e) {
    if (t === "characteristicvaluechanged") {
      const r = this.dataListeners.get(e);
      r && (this.characteristic.removeListener("data", r), this.dataListeners.delete(e));
    }
  }
}
class et {
  noble = null;
  peripheral = null;
  characteristics = {};
  constructor() {
    try {
      this.noble = globalThis.__noble;
    } catch {
      throw new Error(
        "Noble is not installed. Please run: npm install @stoprocent/noble"
      );
    }
  }
  /**
   * Check if Bluetooth is available (Noble is loaded)
   * The powered on state is checked during requestDevice()
   */
  isAvailable() {
    return this.noble !== null;
  }
  /**
   * Scan for and request a Bluetooth printer device
   * Automatically finds devices with MXW01 printer service UUID
   */
  async requestDevice() {
    return new Promise((t, e) => {
      const r = setTimeout(() => {
        this.noble.stopScanning(), e(new Error("Device scan timeout (30s)"));
      }, 3e4), i = (n) => {
        this.noble.stopScanning(), clearTimeout(r), this.peripheral = n, this.noble.removeListener("discover", i), t({
          id: n.id || n.uuid,
          name: n.advertisement.localName || "MXW01 Printer"
        });
      };
      this.noble.on("discover", i);
      const s = () => {
        console.log("Scanning for MXW01 printer..."), this.noble.startScanning(
          [u.PRINTER_SERVICE, u.PRINTER_SERVICE_ALT],
          !1
        );
      };
      if (this.noble.state === "poweredOn")
        s();
      else {
        const n = (c) => {
          c === "poweredOn" && (this.noble.removeListener("stateChange", n), s());
        };
        this.noble.on("stateChange", n);
      }
    });
  }
  /**
   * Connect to a Bluetooth device and get printer service characteristics
   */
  async connect(t) {
    if (!this.peripheral)
      throw new Error("No peripheral found. Call requestDevice() first.");
    try {
      await this.peripheral.connectAsync(), console.log("Connected to peripheral");
      const { characteristics: e } = await this.peripheral.discoverAllServicesAndCharacteristicsAsync();
      if (console.log(`Found ${e.length} characteristics`), this.characteristics.control = e.find(
        (i) => i.uuid === u.CONTROL_SHORT
      ), this.characteristics.notify = e.find(
        (i) => i.uuid === u.NOTIFY_SHORT
      ), this.characteristics.data = e.find(
        (i) => i.uuid === u.DATA_SHORT
      ), console.log("Control:", this.characteristics.control ? "✅" : "❌"), console.log("Notify:", this.characteristics.notify ? "✅" : "❌"), console.log("Data:", this.characteristics.data ? "✅" : "❌"), !this.characteristics.control || !this.characteristics.notify || !this.characteristics.data)
        throw new Error(
          `Missing required characteristics. Found: ${Object.keys(
            this.characteristics
          ).join(", ")}`
        );
      const r = this.peripheral;
      return {
        device: t,
        disconnect: async () => {
          r && r.state === "connected" && (await r.disconnectAsync(), console.log("Disconnected from peripheral")), this.peripheral = null, this.characteristics = {};
        },
        controlCharacteristic: new R(
          this.characteristics.control
        ),
        dataCharacteristic: new R(
          this.characteristics.data
        ),
        notifyCharacteristic: new R(
          this.characteristics.notify
        )
      };
    } catch (e) {
      if (this.peripheral && this.peripheral.state === "connected")
        try {
          await this.peripheral.disconnectAsync();
        } catch (r) {
          console.error("Error disconnecting:", r);
        }
      throw new Error(
        `Failed to connect to device: ${e.message}`
      );
    }
  }
}
export {
  d as Command,
  S as MIN_DATA_BYTES,
  W as MXW01Printer,
  et as NodeBluetoothAdapter,
  y as PRINTER_WIDTH,
  m as PRINTER_WIDTH_BYTES,
  Z as ThermalPrinterClient,
  tt as WebBluetoothAdapter,
  x as encode1bppRow,
  B as prepareImageDataBuffer,
  K as processImageForPrinter
};
